From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 14 May 2024 04:23:20 +0300
Subject: [PATCH] Real custom items

Allows creating real server-side items. Depends on Item Modifiers API to fake the display of the item.

diff --git a/src/main/java/com/destroystokyo/paper/event/executor/asm/SafeClassDefiner.java b/src/main/java/com/destroystokyo/paper/event/executor/asm/SafeClassDefiner.java
index abcc966d8ee01d73c1d1480237ab46fa0ab55fdc..de55e47fda751d09647ac04f51578d76a52a823a 100644
--- a/src/main/java/com/destroystokyo/paper/event/executor/asm/SafeClassDefiner.java
+++ b/src/main/java/com/destroystokyo/paper/event/executor/asm/SafeClassDefiner.java
@@ -8,7 +8,7 @@ import com.google.common.collect.MapMaker;
 import org.jetbrains.annotations.NotNull;
 
 public class SafeClassDefiner implements ClassDefiner {
-    /* default */ static final SafeClassDefiner INSTANCE = new SafeClassDefiner();
+    /* default */ public static final SafeClassDefiner INSTANCE = new SafeClassDefiner(); // Kiterino - Real custom items
 
     private SafeClassDefiner() {}
 
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjector.java b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjector.java
new file mode 100644
index 0000000000000000000000000000000000000000..1888e15525b73b03fd6aa65f6a077fb19cd3b23d
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoDataInjector.java
@@ -0,0 +1,19 @@
+package me.sosedik.kiterino.util;
+
+import me.sosedik.kiterino.world.item.KiterinoNMSItem;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Internal
+public abstract class KiterinoDataInjector {
+
+	static KiterinoDataInjector injector;
+
+	protected abstract @NotNull KiterinoNMSItem modifyNmsItemClass0(@NotNull Class<? extends KiterinoNMSItem> itemClass, @NotNull Object... constructorArgs);
+
+	@ApiStatus.Internal
+	public static @NotNull KiterinoNMSItem modifyNmsItemClass(@NotNull Class<? extends KiterinoNMSItem> itemClass, @NotNull Object... constructorArgs) {
+		return injector.modifyNmsItemClass0(itemClass, constructorArgs);
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/item/KiterinoItem.java b/src/main/java/me/sosedik/kiterino/world/item/KiterinoItem.java
new file mode 100644
index 0000000000000000000000000000000000000000..f6aafa01eb558435e44a8045eb824512ee8d4d80
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/item/KiterinoItem.java
@@ -0,0 +1,54 @@
+package me.sosedik.kiterino.world.item;
+
+import me.sosedik.kiterino.modifier.item.ItemContextBox;
+import me.sosedik.kiterino.util.KiterinoDataInjector;
+import org.bukkit.Material;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+
+/**
+ * Wrapper for simple Kiterino items
+ */
+public interface KiterinoItem extends KiterinoNMSItem {
+
+	/**
+	 * Creates a bare minimum item
+	 *
+	 * @param material material to show to the client
+	 * @param cmd optional CustomModelData to the client
+	 * @param properties item properties
+	 * @return nms wrapper for the item
+	 */
+	static @NotNull Supplier<KiterinoItem> simpleItem(@NotNull Material material, @Nullable Integer cmd, @Nullable UnaryOperator<KiterinoItemProperties> properties) {
+		var props = properties == null ? new KiterinoItemProperties() : properties.apply(new KiterinoItemProperties());
+		return () -> new KiterinoItem() {
+			@Override
+			public void modify(@NotNull ItemContextBox contextBox) {
+				contextBox.setNewMaterial(material);
+				if (cmd != null) {
+					contextBox.getMeta().setCustomModelData(cmd);
+				}
+			}
+
+			@Override
+			public @Nullable KiterinoItemProperties getItemProperties() {
+				return props;
+			}
+		};
+	}
+
+	/**
+	 * Creates item from direct nms implementation
+	 *
+	 * @param itemClass nms item class
+	 * @param constructorArgs values to pass to itemClass's constructor
+	 * @return nms wrapper for the item
+	 */
+	static @NotNull Supplier<KiterinoNMSItem> nmsItem(@NotNull Class<? extends KiterinoNMSItem> itemClass, @NotNull Object... constructorArgs) {
+		return () -> KiterinoDataInjector.modifyNmsItemClass(itemClass, constructorArgs);
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/item/KiterinoItemProperties.java b/src/main/java/me/sosedik/kiterino/world/item/KiterinoItemProperties.java
new file mode 100644
index 0000000000000000000000000000000000000000..6dbd9a775780fdea471c84bc8548596776686fbc
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/item/KiterinoItemProperties.java
@@ -0,0 +1,40 @@
+package me.sosedik.kiterino.world.item;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.Range;
+
+public class KiterinoItemProperties {
+
+	private int maxStackSize = 64;
+	private boolean fireResistant;
+	private @Nullable Float compostChance;
+
+	public int getMaxStackSize() {
+		return this.maxStackSize;
+	}
+
+	public boolean isFireResistant() {
+		return this.fireResistant;
+	}
+
+	public @Nullable Float getCompostChance() {
+		return this.compostChance;
+	}
+
+	public @NotNull KiterinoItemProperties withStackSize(int maxStackSize) {
+		this.maxStackSize = maxStackSize;
+		return this;
+	}
+
+	public @NotNull KiterinoItemProperties fireResistant() {
+		this.fireResistant = true;
+		return this;
+	}
+
+	public @NotNull KiterinoItemProperties compostable(@Range(from = 0, to = 1) float levelIncreaseChance) {
+		this.compostChance = levelIncreaseChance;
+		return this;
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/item/KiterinoNMSItem.java b/src/main/java/me/sosedik/kiterino/world/item/KiterinoNMSItem.java
new file mode 100644
index 0000000000000000000000000000000000000000..f7425ed96d9e2af5734f221aa27f4c123152ca09
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/item/KiterinoNMSItem.java
@@ -0,0 +1,34 @@
+package me.sosedik.kiterino.world.item;
+
+import me.sosedik.kiterino.modifier.item.ItemContextBox;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents Kiterino item
+ * (used for injecting custom items into registry).
+ *
+ * <p>Item behaviours might be extended by implementing following classes:
+ * <ul>
+ * </ul>
+ */
+public interface KiterinoNMSItem {
+
+	/**
+	 * Modifies an item before sending it to the client.
+	 * <p><b>Implementations must override the item type to match one of the vanilla ones.
+	 *
+	 * @param contextBox item data
+	 */
+	void modify(@NotNull ItemContextBox contextBox);
+
+	/**
+	 * Gets the default properties of the item
+	 *
+	 * @return the default properties of the item
+	 */
+	default @Nullable KiterinoItemProperties getItemProperties() {
+		return null;
+	}
+
+}
